KAFKA-6513: Corrected how Converters and HeaderConverters are instantiated and configured

rhauch opened the pull request:
The commits for KIP-145 (KAFKA-5142) changed how the Connect workers instantiate and configure the Converters, and also added the ability to do the same for the new HeaderConverters. However, the last few commits removed the default value for the `converter.type` property for Converters and HeaderConverters, and this broke how the internal converters were being created.

This change corrects the behavior so that the `converter.type` property is always set by the worker (or by the Plugins class), which means the existing Converter implementations will not have to do this. The built-in JsonConverter, ByteArrayConverter, and StringConverter also implement HeaderConverter which implements Configurable, but the Worker and Plugins methods do not yet use the `Configurable.configure(Map)` method and instead still use the `Converter.configure(Map,boolean)`.

Several tests were modified, and a new PluginsTest was added to verify the new behavior in Plugins for instantiating and configuring the Converter and HeaderConverter instances.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-6513: Corrected how Converters and HeaderConverters are instantiated and configured

The commits for KIP-145 (KAFKA-5142) changed how the Connect workers instantiate and configure the Converters, and also added the ability to do the same for the new HeaderConverters. However, the last few commits removed the default value for the `converter.type` property for Converters and HeaderConverters, and this broke how the internal converters were being created.

This change corrects the behavior so that the `converter.type` property is always set by the worker (or by the Plugins class), which means the existing Converter implementations will not have to do this. The built-in JsonConverter, ByteArrayConverter, and StringConverter also implement HeaderConverter which implements Configurable, but the Worker and Plugins methods do not yet use the `Configurable.configure(Map)` method and instead still use the `Converter.configure(Map,boolean)`.

Several tests were modified, and a new PluginsTest was added to verify the new behavior in Plugins for instantiating and configuring the Converter and HeaderConverter instances.

rhauch commented:
@kkonstantine and @ewencp: here's my proposed fix for KAFKA-6513.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/276/

rhauch committed:
KAFKA-6513: Centralized the configuration of the Converter and HeaderConverter within Plugins

Change the Worker and Plugins classes to move all of the logic of configuring the Converter and HeaderConverter instances into the Plugins class. This simplifies the Worker class, centralizes the logic (especially that of adding the `converter.type` configuration property), and makes it easier to test.

rhauch commented:
Added a second commit that moves all of the logic of configuring the `Converter` and `HeaderConverter` instances into the `Plugins` class. This simplifies the `Worker` class, centralizes the logic (especially that of adding the `converter.type` configuration property), and makes the logic easier to test.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/295/

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -260,6 +269,7 @@ private PluginScanResult scanPluginPath(
         return new PluginScanResult(
                 getPluginDesc(reflections, Connector.class, loader),
                 getPluginDesc(reflections, Converter.class, loader),
+                getPluginDesc(reflections, HeaderConverter.class, loader),
```
number of scans & total overall cost is probably getting out of hand now. this works as a fix for now, but we need to fix the fact that every call to `getPluginDesc` is costing a full scan here. /cc @kkonstantine.

rhauch replied:
IIUC, the scanning of the file system occurs in the `Reflections` constructor (via its protected `scan()` method), and all of the results are put into its `Store` instance that is an in-memory cache of the metadata. The `getPluginDesc(...)` method calls `reflections.getSubTypesOf(klass)` that looks in this _in-memory `Store`_ and returns the classes that are subtypes of `klass`. The resulting plugin descriptions are then cached in-memory inside `PluginScanResult` so we can find the classes at a later time.

We instantiate a `DelegatingClassLoader` once for the worker, and that loader instantiates `Reflections` once for the parent classloader (e.g., the classpath) and once for each plugin (that is, for every directory directly inside each plugin path directory. So we do a lot of scanning, but from what I can tell we're only scanning once for the classpath and once for each plugin.

ewencp replied:
I haven't looked at where the costs are, but there was a very noticeable slowdown when we started scanning for more than just connector classes, so I'm pretty sure we're still increasing the cost significantly here. in any case, we really need to look at startup time, with any non-trivial number of connectors it has become very painfully slow.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -185,46 +180,105 @@ public Task newTask(Class<? extends Task> taskClass) {
+     */
+    private <T> T getInstance(AbstractConfig config, String key, Class<T> t) {
+        Class<?> c = config.getClass(key);
+        if (c == null)
```
Is this actually possible? shouldn't a `CLASS` config fail to even parse if it doesn't have a valid value? (which also raises an interesting question re: parsing and ensuring the parsing context is always the right one.)

rhauch replied:
I didn't understand it at first, but this method is ultimately called from the Worker when it sets up its own converters and when it starts up a task to instantiate and initialize the converters, etc. If the worker/connector configuration has a class that is not on the current classloader, then when the `AbstractConfig` is instantiated the value of the `Class<?> c` will be null since the class cannot be found/loaded using the current classloader. That's how we know to go looking in the plugins. Surprising indeed.

rhauch replied:
Something still isn't right, though. Let me work on this some more.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -185,46 +180,105 @@ public Task newTask(Class<? extends Task> taskClass) {
+        if (c == null)
+            return null;
+        Object o = Utils.newInstance(c);
+        if (!t.isInstance(o))
```
Shouldn't this be in the contract of `Utils.newInstance` to not return some other class that doesn't match?

I think this is pulled from `AbstractConfig` which makes sense for consistency, but I don't get why `Utils.newInstance` would ever return a value with an invalid type.

rhauch replied:
The problem is that the class we found in the configuration, `c`, might not actually implement the type `t` that was supplied in the params. We have to do that manually.

BTW, we have to replicate a lot of this logic because `AbstractConfig` has a method to return a _configured_ instance, but no method to return a _non-configured_ instance. The proper way to address this is to add that method to `AbstractConfig`, but that requires a KIP, right?

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -185,46 +180,105 @@ public Task newTask(Class<? extends Task> taskClass) {
+     * @return the instantiated and configured {@link Converter}; never null
+     * @throws ConnectException if the {@link Converter} implementation class could not be found
+     */
+    public Converter newConverter(AbstractConfig config, String classPropertyName, boolean isKeyConverter) {
```
by no means a blocker (especially given public interfaces relying on this same problem), but i would not be opposed to making internal APIs not rely on these booleans -- clear enums, with properly handled default cases (presumably throwing exceptions) would be better. we tend to let things like this leak into internal (and subsequently external) APIs, but I think it is probably always a mistake -- unclear from the caller standpoint and not extensible.

rhauch replied:
Yeah, that's a good point. I'll try to address.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -185,46 +180,105 @@ public Task newTask(Class<? extends Task> taskClass) {
+     */
+    public HeaderConverter newHeaderConverter(AbstractConfig config, String classPropertyName) {
+        // First see if we can instantiate the instance using the current classloader ...
+        HeaderConverter plugin = getInstance(config, classPropertyName, HeaderConverter.class);
```
The code isn't huge, so by no means a blocker, but @kkonstantine pointed out that the entire block before configuration (and most of config modulo the conditional) is identical between header and "normal" converters. The main difference in the header blocks are just the class referenced (`Converter.class` vs `HeaderConverter.class`).

Consolidation would be nice if it's easy to do, but at the same time I'd rather get a fix to the immediate problem in, so definitely wouldn't block on saving 10 lines of duplicated code.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -185,46 +180,105 @@ public Task newTask(Class<? extends Task> taskClass) {
+            plugin = newPlugin(klass);
         }
-        return config != null ? newConfiguredPlugin(config, klass) : newPlugin(klass);
+        if (plugin instanceof Configurable) {
```
This still doesn't seem correct. We're only invoking configuration if the plugin implements `Configurable` afaict. This does not work for `Converter`s implemented against the new API and assuming the "forward" configuration. We *must* always invoke the "old" `configure(Map, boolean)`, and only invoke the `Configurable` version as a backup. Possibly it would make sense to indicate on the `HeaderConverter` that the `Configurable` methods should be idempotent if we need to be able to implement both.

Not sure if we can test this easily with unit tests, but I think we might want a plain old `Converter` (that does not implement `HeaderConverter`) in tests to validate compatibility... but it's possible we'd need either integration or system tests to properly validate.

ewencp replied:
Another way to put this is that maybe we should make sure our built-in converters can handle calls to both `configure(Map, boolean isKey)` followed by `configure(Map)` with the `TYPE_CONFIG`. then, only things that are `Configurable` see the second one. old implementations wouldn't see it as they would not have implemented `Configurable` (except in unusual circumstances). New implementations could be warned by docs in `HeaderConverer` that they should take care to handle that sequence of calls.

rhauch committed:
KAFKA-6513: Corrected the logic that instantiates new converters and header converters

The `Plugins` class now encapsulates the instantiation and configuration of the Converter and HeaderConverter instances. However, when the `Worker` creates the tasks it uses the same logic as before: look first in the connector configuration and use the current classloader (to prefer the connector plugin’s implementations but still find them in any other plugins or the classpath), and if no such configuration is found then use the worker’s configuration.

rhauch commented:
@ewencp, hopefully addressed your concerns about the bad logic; did not address your concerns about scanning because I still think we're only scanning once.

The previous commit did indeed use incorrect logic when trying to instantiate the `Converter` and `HeaderConverter` classes. The `Plugins` class now encapsulates the instantiation and configuration of the `Converter` and `HeaderConverter instances`. However, when the `Worker` creates the tasks it uses the same logic as before: look first in the connector configuration and use the current classloader (to prefer the connector plugin’s implementations but otherwise still find them in any other plugins or the classpath), and if no such configuration is found then use the worker’s configuration (using the normal plugin mechanism, which ends with the classpath)

kkonstantine reviewed (approved):
I'm signing off on this PR due to its urgency (current connect worker configurations are broken since `converter.type` was introduced). 

However there's a lot of redundancy between `newConverter` and `newHeaderConverter` methods, which probably could be addressed in a follow up.

rhauch commented:
@kkonstantine wrote:
> However there's a lot of redundancy between newConverter and newHeaderConverter methods, which probably could be addressed in a follow up.

I agree. I tried to think of a non-complex way to eliminate the redundancy, but the differences are deep within the logic.  This will be very trivial to refactor with Java 8, so I'd prefer to wait. I will log an AK issue and target to 2.0.

kkonstantine commented:
@rhauch Passing the class object for `Converter.class` or `HeaderConverter.class` would suffice more the most part I believe. `Java >= 8` will help

ewencp reviewed (approved):
LGTM, merging to trunk and 1.1
